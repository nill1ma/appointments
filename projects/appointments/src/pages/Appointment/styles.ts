import styled from "styled-components";

export const Container = styled.div`
  display: flex;
  flex-direction: column;
`;

export const CardsContainer = styled.div`
  display: flex;
  flex-wrap: wrap;
  width: 100%;
`;
