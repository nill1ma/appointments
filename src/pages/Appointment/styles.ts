import styled from "styled-components";

type CardsContainerProps = {
	hasDetail?: boolean;
	isOpened?: boolean;
};

export const Container = styled.div`
	display: flex;
	flex-direction: column;
	width: 100vw;
`;

export const CardsContainer = styled.div<CardsContainerProps>`
	width: ${({ hasDetail }) => (hasDetail ? `50%` : `100%`)};
	display: flex;
	flex-wrap: wrap;
	border-right: 1px solid #2c2c2c;
	overflow-y: scroll;
	::-webkit-scrollbar {
		width: 5px;
	}

	/* Track */
	::-webkit-scrollbar-track {
		-webkit-box-shadow: inset 0 0 6px #2c2c2c;
	}
	/* Handle */
	::-webkit-scrollbar-thumb {
		background: #363636;
		-webkit-box-shadow: inset 0 0 6px #2c2c2c;
	}
	::-webkit-scrollbar-thumb:window-inactive {
		background: #363636;
	}
`;

export const Content = styled.div<CardsContainerProps>`
	width: 100%;
	display: flex;
	height: ${({ isOpened }) => (isOpened ? "75vh" : "100vh")};
`;
