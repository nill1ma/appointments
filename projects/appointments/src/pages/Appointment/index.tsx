import React, { useState } from "react";
import Button from "../../components/Button";
import Card from "../../components/Card";
import Header from "../../components/Header";
import { IAppointment } from "../../models/environment";
import { CardsContainer, Container } from "./styles";
import { v4 as uuidv4 } from "uuid";
import { getAppointment, saveAppointment } from "../../services/appointments";
import CreateAppointment from "../../components/CreateAppointment";

export default function Appointment() {
  const [open, setOpen] = useState<boolean>(false);
  const [appointments, setAppointments] = useState<IAppointment[]>(
    getAppointment()
  );

  const save = (appointment: IAppointment) => {
    const app = { ...appointment, id: uuidv4() };
    console.log(new Date(app.date_start), `app`);
    setAppointments([...appointments, app]);
    saveAppointment(app);
  };

  const handleOpenState = () => setOpen(!open);

  return (
    <Container>
      <div>{open && <CreateAppointment open={open} action={save} />}</div>
      <CardsContainer>
        {appointments.map((appointment: IAppointment) => {
          return <Card key={appointment.id} item={appointment} />;
        })}
      </CardsContainer>
      <Button
        bottom={`5%`}
        self_alignment={`flex-end`}
        position={`fixed`}
        label={open ? `Keep it hidden` : `Create an Environment`}
        action={handleOpenState}
      />
    </Container>
  );
}
