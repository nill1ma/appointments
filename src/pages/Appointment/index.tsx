import { useState } from "react";
import { v4 as uuidv4 } from "uuid";
import Button from "../../components/Button";
import Card from "../../components/Card";
import CardDetail from "../../components/CardDetail";
import CreateAppointment from "../../components/CreateAppointment";
import Filters from "../../components/Filters";
import { IAppointment } from "../../models/appointments";
import { References } from "../../models/filters-references";
import { getAppointments, saveAppointment } from "../../services/appointments";
import { CardsContainer, Container, Content } from "./styles";

export default function Appointment() {
	const [isOpened, setIsOpened] = useState<boolean>(false);
	const [appointments, setAppointments] = useState<IAppointment[]>(
		getAppointments()
	);

	const [detail, setDetail] = useState<IAppointment>();

	const save = (appointment: IAppointment) => {
		const app: IAppointment = {
			...appointment,
			id: uuidv4(),
			start: `${appointment.start}:00`,
			end: `${appointment.end}:59`,
		};

		setAppointments([...appointments, app]);
		saveAppointment(app);
	};

	const handleOpenedState = () => setIsOpened(!isOpened);

	const filterAppointments = (reference: References) => {
		const allAppointments: IAppointment[] = getAppointments();
		const filteredAppointments = allAppointments.filter((app: IAppointment) => {
			const { type, value } = reference;
			const date = new Date(app.start);
			// const result = compareReferences(type, d, String(value));
			const result = date.getFullYear() === reference.value;
			return result;
		});
		setAppointments([...filteredAppointments]);
	};

	const compareReferences = (type: string, filter: Date, reference: string) => {
		switch (type) {
			case "MONTH":
				return `${filter.getMonth() + 1}` === reference;
			default:
				return String(filter.getFullYear()) === reference;
		}
	};

	return (
		<Container>
			<>
				{isOpened ? (
					<CreateAppointment isOpened={isOpened} action={save} />
				) : (
					<Filters handleCurrentPeriod={filterAppointments} />
				)}
			</>
			<Content>
				<CardsContainer hasDetail={detail !== undefined}>
					{appointments.map((appointment: IAppointment) => {
						return (
							<Card
								key={appointment.id}
								item={appointment}
								setDetail={setDetail}
							/>
						);
					})}
				</CardsContainer>
				{detail && <CardDetail detail={detail} setDetail={setDetail} />}
			</Content>
			<Button
				bottom={`5%`}
				right={`1%`}
				self_alignment={`flex-end`}
				position={`fixed`}
				label={isOpened ? `Keep it hidden` : `Create an Environment`}
				action={handleOpenedState}
			/>
		</Container>
	);
}
