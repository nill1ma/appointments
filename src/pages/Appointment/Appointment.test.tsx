import "@testing-library/jest-dom";
import { fireEvent, render, screen } from "@testing-library/react";
import Appointment from ".";

describe("Appointment", () => {
	it("Should return defined component", () => {
		const utils = render(<Appointment />);
		expect(utils).toBeDefined();
	});

	it("Should load Create Appointment component when click in Create an Environment", () => {
		render(<Appointment />);
		const button = screen.getByRole("button");
		fireEvent.click(button);
		const placeholder = screen.getAllByPlaceholderText("Type a Description");
		expect(placeholder).toMatchSnapshot();
	});

	it("Should make sure that there is the correct text on button before click on Create an Environment", () => {
		render(<Appointment />);
		const placeholder = screen.getAllByText("Create an Environment");
		expect(placeholder).toMatchSnapshot();
	});

	it("Should make sure that there is the correct text on button after click on Create an Environment", () => {
		render(<Appointment />);
		const button = screen.getByRole("button");
		fireEvent.click(button);
		const placeholderButton = screen.getAllByText("Keep it hidden");
		expect(placeholderButton).toMatchSnapshot();
		fireEvent.click(button);
		const placeholderInitialState = screen.getAllByText(
			"Create an Environment"
		);
		expect(placeholderInitialState).toMatchSnapshot();
	});

	it("Should make sure that there is Save button after click on Create an Environment", () => {
		render(<Appointment />);
		const button = screen.getByRole("button");
		fireEvent.click(button);
		const placeholderInitialState = screen.getAllByText("save");
		expect(placeholderInitialState).toMatchSnapshot();
	});
});
