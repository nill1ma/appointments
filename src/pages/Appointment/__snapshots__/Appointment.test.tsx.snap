// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Appointment Should load Create Appointment component when click in Create an Environment 1`] = `
Array [
  <input
    id="description"
    name="description"
    placeholder="Type a Description"
    type="text"
  />,
]
`;

exports[`Appointment Should make sure that there is Save button after click on Create an Environment 1`] = `
Array [
  <button
    class="sc-bdvvtL ignKBR"
  >
    save
  </button>,
]
`;

exports[`Appointment Should make sure that there is the correct text on button after click on Create an Environment 1`] = `
Array [
  <button
    class="sc-bdvvtL lhuVmG"
  >
    Keep it hidden
  </button>,
]
`;

exports[`Appointment Should make sure that there is the correct text on button after click on Create an Environment 2`] = `
Array [
  <button
    class="sc-bdvvtL lhuVmG"
  >
    Create an Environment
  </button>,
]
`;

exports[`Appointment Should make sure that there is the correct text on button before click on Create an Environment 1`] = `
Array [
  <button
    class="sc-bdvvtL lhuVmG"
  >
    Create an Environment
  </button>,
]
`;
